Improve man page

mariobl opened the pull request:
Hello Graeme,

I've extended the man page a bit:

First, I've merged the output of `tanglet` --help and `tanglet --version` into the *roff source. But for the audience of such a man page, this isn't very helpful yet. IMHO, people who read the man page of a GUI application are advanced users or developers, so I think the Qt specific options (with `--help-all`) shouldn't be omitted.

Moreover, I've tweaked the markup a bit and added a bugreport section and a SEE ALSO section with a Wikipedia link to Boggle.

gottcode commented:
This is a good start, but I have some thoughts before I merge this.

First, you should probably make the ```--version``` option also mention that ```-v``` does the same thing, since you're being so thorough documenting all the options.

Secondly, I hesitate to include the ```--qmljsdebugger``` option, because to me including it implies that it does something with Tanglet. But since Tanglet does not have nor use any QML or JavaScript, the option is useless. In fact, when I try to pass it to Tanglet, it actually says the option is unknown despite ```--help-all``` saying it is a valid option!

Finally, I don't have most of those options when I run ```--help-all```; the options stop with ```--session```. Do you see them when you run ```--help-all```?

mariobl commented:
> This is a good start, but I have some thoughts before I merge this.
> 
> First, you should probably make the `--version` option also mention that `-v` does the same thing, since you're being so thorough documenting all the options.
>
OK, added.
> Secondly, I hesitate to include the `--qmljsdebugger` option, because to me including it implies that it does something with Tanglet. But since Tanglet does not have nor use any QML or JavaScript, the option is useless. In fact, when I try to pass it to Tanglet, it actually says the option is unknown despite `--help-all` saying it is a valid option!
> 
Then let's omit this option. Assuming that these options are introduced by some Qt mechanism, it is strange to have such an entry without having the corresponding code anywhere.

> Finally, I don't have most of those options when I run `--help-all`; the options stop with `--session`. Do you see them when you run `--help-all`?
>
```
[mariob@t450 doc]$ tanglet --help-all
Usage: tanglet [options] [file]
Buchstabenrätsel

Options:
  -h, --help                           Displays help on commandline options.
  --help-all                           Displays help including Qt specific
                                       options.
  -v, --version                        Displays version information.
  --qmljsdebugger <value>              Activates the QML/JS debugger with a
                                       specified port. The value must be of
                                       format port:1234[,block]. "block" makes
                                       the application wait for a connection.
  --platform <platformName[:options]>  QPA plugin. See QGuiApplication
                                       documentation for available options for
                                       each plugin.
  --platformpluginpath <path>          Path to the platform plugins.
  --platformtheme <theme>              Platform theme.
  --plugin <plugin>                    Additional plugins to load, can be
                                       specified multiple times.
  --qwindowgeometry <geometry>         Window geometry for the main window,
                                       using the X11-syntax, like 100x100+50+50.
  --qwindowicon <icon>                 Default window icon.
  --qwindowtitle <title>               Title of the first window.
  --reverse                            Sets the application's layout direction
                                       to Qt::RightToLeft (debugging helper).
  --session <session>                  Restores the application from an earlier
                                       session.
  --display <display>                  Display name, overrides $DISPLAY.
  --name <name>                        Instance name according to ICCCM
                                       4.1.2.5.
  --nograb                             Disable mouse grabbing (useful in
                                       debuggers).
  --dograb                             Force mouse grabbing (even when running
                                       in a debugger).
  --visual <id>                        ID of the X11 Visual to use.
  --geometry <geometry>                Alias for --windowgeometry.
  --icon <icon>                        Alias for --windowicon.
  --title <title>                      Alias for --windowtitle.

Arguments:
  file                                 Spieldatei zum Spielen.
```
Do you already use Wayland? Maybe the X11-related options don't appear then.

> **Edit:** Another minor note is that the copyright of Tanglet is actually **2009**-2021, not 2014-2021.
Your current man page is copyright 2014 ;) Updated to »2009-2021«.

Moreover, I've added a FILES section with some short descriptions. Do you think it could be useful to have a tanglet(5) man page which describes the files more detailed...?

gottcode commented:
> Do you already use Wayland? Maybe the X11-related options don't appear then.

Hmm. I run X11 from ```startx```, not a display manager. Maybe that's why? I would like to use Wayland at some point, but there are too many bugs for me right now.

> Moreover, I've added a FILES section with some short descriptions. Do you think it could be useful to have a tanglet(5) man page which describes the files more detailed...?

I'm not sold on the FILES section. The fact that the data file is GZip compressed is only an implementation detail that I could change in the future if I decide to use a different algorithm.

I also don't intend for users to modify the data files. I feel free to completely change the format, and have done so in the past. Same goes for the configuration file. Yes, they are both INI files, so users *can* change them, but that again is only an implementation detail. Also, the specific settings have changed from version to version (for instance, I completely replaced how the scores are stored with version 1.6.0).

Because of that, I don't think there should be a tanglet(5) detailing the files--they aren't meant to be modified by hand, and I don't want to commit to any specification.

gottcode commented:
I have a few more thoughts:

Every manpage source I have seen puts quotes around the text after ```.SH```, and you did for some but not all of the sections you added. For consistency, I would think they should all be quoted.

I think that ```GENERAL OPTIONS``` should just be ```OPTIONS```, to match other manpages.

Since there is a link to Boggle in the GUI already, I think the ```SEE ALSO``` is unnecessary. Tanglet is, after all, a GUI program. A manpage is supposed to be a quick reference for how to *use* a program. Also, looking at the generally expected manpage layout definition in ```man-pages(7)```, the ```SEE ALSO``` section refers to other manpages.

Thanks for working on this!

mariobl commented:
> Every manpage source I have seen puts quotes around the text after `.SH`, and you did for some but not all of the sections you added. For consistency, I would think they should all be quoted.
>

OK, although it doesn't have any effect in the rendered man page.

> I think that `GENERAL OPTIONS` should just be `OPTIONS`, to match other manpages.
> 

OK, changed.

> Since there is a link to Boggle in the GUI already, I think the `SEE ALSO` is unnecessary. Tanglet is, after all, a GUI program. A manpage is supposed to be a quick reference for how to _use_ a program. Also, looking at the generally expected manpage layout definition in `man-pages(7)`, the `SEE ALSO` section refers to other manpages.
> 
OK, removed the section. Moreover, I've removed the FILES section. Indeed, Tanglet as a GUI program doesn't need any links to other GUI program's man pages.

Well, Tanglet has a man page as a short reference to its command line options. But what is missing here, is a real user manual, which can be opened directly via the menu bar (Help > Contents …, as known from other user manuals) or using a keyboard shortcut (usually F1). If this is interesting for you, we can discuss this in another issue. I would write a user manual, which could be displayed then in Qtassistant, which is available from all platforms Qt supports, *nix, Windows, macOS.

gottcode commented:
This looks good! I am going to merge the updates to the manpage.

I would be quite happy to add a user manual! I would probably match the layout of the KDE games manuals, if you don't have any objections.

mariobl commented:
> I would be quite happy to add a user manual! I would probably match the layout of the KDE games manuals, if you don't have any objections.

Yes, it would match the layout, but because Tanglet is rather OS-agnostic than KDE-related, I proposed already to use a framework for displaying it in Qtassistant. The manual itself will then be shipped in HTML; do you have any preferences regarding the source format? Maybe you don't know why I ask this: Because Tanglet is already translated into many languages, the user manual should also be localized. HTML is rather difficult to translate – means: using .po files. Translating a plain HTML file is no acceptable option. I would prefer Asciidoc (which needs Ruby) or Sphinx (needs Python). Asciidoc(tor) is perfect for the [Po4a](https://po4a.org/) toolkit, [Python Sphinx](https://www.sphinx-doc.org/en/master/#) has a builtin mechanism for translations using .po files.

gottcode commented:
> Yes, it would match the layout, but because Tanglet is rather OS-agnostic than KDE-related, I proposed already to use a framework for displaying it in Qtassistant. The manual itself will then be shipped in HTML; do you have any preferences regarding the source format? Maybe you don't know why I ask this: Because Tanglet is already translated into many languages, the user manual should also be localized. HTML is rather difficult to translate – means: using .po files. Translating a plain HTML file is no acceptable option. I would prefer Asciidoc (which needs Ruby) or Sphinx (needs Python). Asciidoc(tor) is perfect for the [Po4a](https://po4a.org/) toolkit, [Python Sphinx](https://www.sphinx-doc.org/en/master/#) has a builtin mechanism for translations using .po files.

I'm aware that the manual would need to be translated. :smile: I don't have any preferences regarding the source format. I'm only passingly familiar with AsciiDoc and I've never used Sphinx, so I will need to play around with both for a bit to determine which will be easier to work with and integrate into my workflow. Without having really used either I would think that Sphinx would be the simpler solution since it handles translations, and AsciiDoc would require something like po4a.

gottcode commented:
Back to the manpage:

There actually is a ```SEE ALSO``` manpage that Tanglet could reference: ```qt5options(7)```, which could then be used to replace the entire ```QT-SPECIFIC OPTIONS``` section. That is how KDE programs reference all of the shared options for Qt programs. The ```qtoptions(7)``` manpage actually comes from KDE, not Qt, but I don't expect there would be that many players who have my Qt-based programs installed and not a single KDE program installed.

The options are the same between Qt 5 and Qt 6, and I expect at some point there will be a qt6options manpage that could be used instead.

mariobl commented:
> Without having really used either I would think that Sphinx would be the simpler solution since it handles translations, and AsciiDoc would require something like po4a.

It wouldn't make a big difference; Po4a handles both traditional Asciidoc and Asciidoctor perfectly, and it only needs a few Perl modules to work. However, if you like to use Sphinx, [Glabels-qt](https://github.com/jimevins/glabels-qt/tree/UserManual/user-docs/manual) is an example for a working framework. I've initiated this last year, and the import of the previous user manual from the GTK version of Glabels is still on my TODO list. But it is a lot of work – and Python Sphinx is not free from pitfalls, as is RestructuredText. Maybe it is easier to write a small but complete user manual for Tanglet first and get it working, and then continue on Glabels-qt with the experience gained.

> There actually is a SEE ALSO manpage that Tanglet could reference: qt5options(7), which could then be used to replace the entire QT-SPECIFIC OPTIONS section. That is how KDE programs reference all of the shared options for Qt programs.

Sounds good, this would solve the problem with the differences in --help-all output on your and my machine.

gottcode commented:
> It wouldn't make a big difference; Po4a handles both traditional Asciidoc and Asciidoctor quite good, and it only needs a few Perl modules to work. However, if you like to use Sphinx, [Glabels-qt](https://github.com/jimevins/glabels-qt/tree/UserManual/user-docs/manual) is an example for a working framework. I've initiated this last year, and the import of the previous user manual from the GTK version of Glabels is still on my TODO list. But it is a lot of work – and Python Sphinx is not free from pitfalls, as is RestructuredText. Maybe it is easier to write a small but complete user manual for Tanglet first and get it working, and then continue on Glabels-qt with the experience gained.

After playing with both AsciiDoc + Po4a and Sphinx, I have settled on having the user manual be a bunch of HTML files created and translated by Sphinx (instead of QtHelp files needing Qt Assistant). I have a preliminary version in the ```user-manual``` branch that you can work with and give me feedback.

AsciiDoc is much simpler, but it required messy hacks to create a two-column layout the way that Sphinx has by default. Plus the .po files created by Po4a included newlines at the end of the regular paragraphs, and I know from past experience that not all translators will remember to keep those, which would mean more headache when updating the files.

mariobl commented:
> AsciiDoc is much simpler, but it required messy hacks to create a two-column layout the way that Sphinx has by default.

To mention, Qtassistant has this navigation sidebar by default ;) But let's go ahead with Sphinx, to view the manual in $BROWSER, or its pendants on non-*nix target platforms.

gottcode commented:
> To mention, Qtassistant has this navigation sidebar by default ;) But let's go ahead with Sphinx, to view the manual in $BROWSER, or its pendants on non-*nix target platforms.

True, but that requires dragging the Qt Assistant program along on Windows with each program I add a user manual for. ;) And the files are a lot bigger than the plain HTML files--over 3MB overall versus around 300KB! Sure, each language was only around 160KB, but when you compare that to around 20KB for plain HTML, it does add up.

I also had an issue where the Qt Assistant didn't want to show the correct manual for a language if I had already opened it once before with a different language. It still thought the previous language was "registered" for the program, so I would have to find a way to add the language name to the QtHelp string... at which point I decided it wasn't worth it because of the other issues. :)

You can see some artifacts from when I was working with Qt Assistant that I forgot to clean up.
